components/merkledb/src/backends/temporarydb.rs

```diff
@@ -91,22 +84,137 @@ impl Database for TemporaryDB {
+                let mut buffer: [u8; 1_024] = [0; 1024];
+                unsafe {
+                    ptr::copy(id_bytes.as_ptr(), buffer.as_mut_ptr(), ID_SIZE);
+                }
+
+                for (key, change) in changes.into_data() {
+                    let buffer_size = key.len() + ID_SIZE;
+                    unsafe {
+                        ptr::copy(key.as_ptr(), buffer.as_mut_ptr().add(ID_SIZE), key.len());
+                    }
+                    match change {
+                        Change::Put(value) => {
+                            collection.insert(buffer[..buffer_size].into(), value)
+                        }
+                        Change::Delete => collection.remove(&buffer[..buffer_size]),
+                    };
+                }
+            } else {
+                // Write changes to the column family as-is.
+                for (key, change) in changes.into_data() {
+                    match change {
+                        Change::Put(value) => collection.insert(key, value),
+                        Change::Delete => collection.remove(&key),
+                    };
+                }
+            }
+        }
+        Ok(())
     }
 
     fn merge_sync(&self, patch: Patch) -> Result<()> {
-        self.inner.merge_sync(patch)
+        self.merge(patch)
+    }
+}
+
+impl<'a> DbIterator for TemporaryDBIterator<'a> {
+    fn next(&mut self) -> Option<(&[u8], &[u8])> {
+        if self.ended {
+            return None;
+        }
+
+        let (key, value) = self.iter.next()?;
+
+        if let Some(ref prefix) = self.prefix {
+            if &key[..ID_SIZE] != prefix {
+                self.ended = true;
+                return None;
+            }
+        }
+
+        let key = if self.prefix.is_some() {
+            &key[ID_SIZE..]
+        } else {
+            &key[..]
+        };
+
+        Some((key, value))
+    }
+
+    fn peek(&mut self) -> Option<(&[u8], &[u8])> {
+        if self.ended {
+            return None;
+        }
+
+        let (key, value) = self.iter.peek()?;
+        let key = if let Some(prefix) = self.prefix {
+            if key[..ID_SIZE] != prefix {
+                self.ended = true;
+                return None;
+            }
+            &key[ID_SIZE..]
+        } else {
+            &key[..]
+        };
+
+        Some((key, value))
     }
 }
 
 impl Snapshot for TemporarySnapshot {
     fn get(&self, name: &ResolvedAddress, key: &[u8]) -> Option<Vec<u8>> {
-        self.snapshot.get(name, key)
+        let collection = self.snapshot.get(&name.name)?;
+        collection.get(name.keyed(key).as_ref()).cloned()
     }
 
     fn iter(&self, name: &ResolvedAddress, from: &[u8]) -> Iter<'_> {
-        self.snapshot.iter(name, from)
+        let collection = self
+            .snapshot
+            .get(&name.name)
+            .or_else(|| self.snapshot.get(DEFAULT_COLLECTION))
+            .unwrap();
+        let from = name.keyed(from).to_vec();
```

Comment:
[`range` is parametric](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) by the key parameter; AFAIK, it's precisely to avoid unnecessary heap allocations, such as this one.

Reply:
... by the key parameter of the collection. A key of the collection has a type `Vec<u8>` and therefore we should use `Vec<u8>` in a `range`, shouldn't we ?